include/glow/Optimizer/Optimizer.h

```diff
@@ -38,14 +39,18 @@ void optimize(Function *F, const CompilationOptions &opts);
 void optimize(Function *F, CompilationMode mode);
 
 /// Lower the high-level neural network nodes found in \p F into low-level
-/// linear algebra operators. If \p B is not a nullptr then it can prevent
-/// lowering of a node via \ref Backend::shouldLower(); otherwise everything
-/// will be lowered. If \p loweredMap is not a nullptr, then \p loweredMap will
-/// contain a mapping from output names of the nodes found and lowered in \p F
-/// to the output names of the nodes they were lowered from along with the
-/// NodeKind. \p doNotLowerKinds is a set of NodeKinds which represents all
-/// Nodes that should not be lowered.
-void lower(Function *F, LoweredInfoMap *loweredMap, const Backend *B = nullptr,
+/// linear algebra operators. If \p backend is not a nullptr then it can prevent
+/// lowering of a node via \ref Backend::shouldLower(); likewise if \p
+/// backendKind is not null then Backend::shouldLowerStatic can prevent
+/// lowering; otherwise everything will be lowered. At most of one of backend
+/// and backendKind should be non-null. If \p loweredMap is not a nullptr, then
+/// \p loweredMap will contain a mapping from output names of the nodes found
+/// and lowered in \p F to the output names of the nodes they were lowered from
+/// along with the NodeKind. \p doNotLowerKinds is a set of NodeKinds which
+/// represents all Nodes that should not be lowered.
+void lower(Function *F, LoweredInfoMap *loweredMap,
```

Comment:
The lower method now has 5 parameters. Do we need all of the parameters? Can we refactor and clean things up here?

Reply:
Yeah it would be good to refactor but I'm not sure what's the best thing to do.

It would be good if we could just use `BackendKind` in `lower` and then call the `shouldLowerStatic()` but we have some one off Backends in unit tests that don't have a BackendKind and implement their own `shouldLower()` methods for example `MockBackendUnloweredFC` so that won't work for all cases.

We could create two lower methods, one that uses backend for unit tests and one that uses backend kind for everything else but I'm not sure that's better. What do you think?

Reply:
I prefer the first one. For `MockBackendUnloweredFC`, can we add a `shouldLower()` method simply return true or false?